src/Solvers/matrix_iterative_solver.jl

```diff
@@ -0,0 +1,273 @@
+
+    return matrix_constructors, diag, N
+end
+
+@kernel function _initialize_variable_diagonal!(diag, D, N)  
+    i, j, k = @index(Global, NTuple)
+    t  = i + N[1] * (j - 1 + N[2] * (k - 1))
+    diag[t] = D[i, j, k]
+end
+
+function fill_core_matrix!(coeff_d, coeff_x, coeff_y, coeff_z, Ax, Ay, Az, C, N, dims)
+    Nx, Ny, Nz = N
+    for k = 1:Nz, j = 1:Ny, i = 1:Nx
+        t          = i +  Nx * (j - 1 + Ny * (k - 1))
+        coeff_d[t] = C[i, j, k]
+    end
+    if dims[1]
+        for k = 1:Nz, j = 1:Ny, i = 1:Nx-1
+            t             = i +  Nx * (j - 1 + Ny * (k - 1))
+            coeff_x[t]    = Ax[i+1, j, k] 
+            coeff_d[t]   -= coeff_x[t]
+            coeff_d[t+1] -= coeff_x[t]
+        end
+    end
+    if dims[2]
+        for k = 1:Nz, j = 1:Ny-1, i = 1:Nx
+            t              = i +  Nx * (j - 1 + Ny * (k - 1))
+            coeff_y[t]     = Ay[i, j+1, k] 
+            coeff_d[t]    -= coeff_y[t] 
+            coeff_d[t+Nx] -= coeff_y[t]
+        end
+    end
+    if dims[3]
+        for k = 1:Nz-1, j = 1:Ny, i = 1:Nx
+            t                 = i +  Nx * (j - 1 + Ny * (k - 1))
+            coeff_z[t]        = Az[i, j, k+1] 
+            coeff_d[t]       -= coeff_z[t]
+            coeff_d[t+Nx*Ny] -= coeff_z[t]
+        end
+    end
+end
+
+ @inline fill_boundaries_x!(coeff_d, coeff_bound_x, Ax, N, ::Type{Bounded}) = nothing
+ @inline fill_boundaries_x!(coeff_d, coeff_bound_x, Ax, N, ::Type{Flat})    = nothing
+function fill_boundaries_x!(coeff_d, coeff_bound_x, Ax, N, ::Type{Periodic})
+    Nx, Ny, Nz = N
+    for k = 1:Nz, j = 1:Ny
+        tₘ = 1  + Nx * (j - 1 + Ny * (k - 1))
+        tₚ = Nx + Nx * (j - 1 + Ny * (k - 1))
+        coeff_bound_x[tₘ] = Ax[1, j, k]
+        coeff_d[tₘ]      -= coeff_bound_x[tₘ]
+        coeff_d[tₚ]      -= coeff_bound_x[tₘ]
+    end
+end
+
+ @inline fill_boundaries_y!(coeff_d, coeff_bound_y, Ay, N, ::Type{Bounded}) = nothing
+ @inline fill_boundaries_y!(coeff_d, coeff_bound_y, Ay, N, ::Type{Flat})    = nothing
+function fill_boundaries_y!(coeff_d, coeff_bound_y, Ay, N, ::Type{Periodic})
+    Nx, Ny, Nz = N
+
+    for k = 1:Nz, i = 1:Nx
+        tₘ = i + Nx * (1 - 1 + Ny * (k - 1))
+        tₚ = i + Nx * (Ny- 1 + Ny * (k - 1))
+        coeff_bound_y[tₘ] = Ay[i, 1, k]
+        coeff_d[tₘ]      -= coeff_bound_y[tₘ]
+        coeff_d[tₚ]      -= coeff_bound_y[tₘ]
+    end
+end
+    
+ @inline fill_boundaries_z!(coeff_d, coeff_bound_z, Az, N, ::Type{Bounded}) = nothing 
+ @inline fill_boundaries_z!(coeff_d, coeff_bound_z, Az, N, ::Type{Flat})    = nothing
+function fill_boundaries_z!(coeff_d, coeff_bound_z, Az, N, ::Type{Periodic})
+    Nx, Ny, Nz = N
+    for j = 1:Ny, i = 1:Nx
+        tₘ = i + Nx * (j - 1 + Ny * (1 - 1))
+        tₚ = i + Nx * (j - 1 + Ny * (Nz- 1))
+        coeff_bound_z[tₘ] = Az[i, j, 1]
+        coeff_d[tₘ]      -= coeff_bound_z[tₘ]
+        coeff_d[tₚ]      -= coeff_bound_z[tₘ]
+    end
+end
```

Comment:
I don't have any idea what's going on in the above code. Would comments be appropriate here?

Reply:
Ok I'll comment a bit